### PR TITLE
[igemm] Adding flexible layout support

### PR DESCRIPTION
**Pull request by jerryyin**
* Modified tensor_api and descriptor for layout updates
* Updated conv_driver to set tensor layouts
* Updated problem_description to avoid key collision
* Enabled implicit_gemm to get layouts from tensors

cc @whchung @zjing14 @asroy 

--------------

Note: This is going to be the first PR in the series to support igemm kernel generation. This PR aims to pave the road for introducing igemm runtime generated kernels. I mark the PR for draft pull request so that we start with the API design and then the implementation details.

~~Specifically, I want to get feedback on the `miopenSet4dTensorDescriptorLayout` API.~~

~~- As a reference, Cuda's related solution is a `cudnnTensorFormat_t`, which includes either `NCHW` or `NHWC`.~~
~~This is less desirable because igemm kernels supports way more layout than the two. However, it is also valid point if we want to hide this capability from user. If we follow their approach, it will require a number of new APIs on setting and getting the descriptor.~~
~~A compromise is to still use the current API, but invent a new enum: `miopenTensorFormat_t`. Since formats needs to be able to convey the difference between input/filter/output tensors, this will ended up in a huge array of enums one day. This does not seem to be a clean solution either.~~

### CODE REVIEW

**Comment by jerryyin** [comment]
FYI I cannot reproduce the failures from CI, [here](http://micimaster.amd.com/blue/rest/organizations/jenkins/pipelines/MLLibs/pipelines/MIOpen/branches/zyin-layouts/runs/1/nodes/459/steps/501/log/?start=0). Please let me know if there is better ways to capture those issues.

Build command
`CXX=/opt/rocm/llvm/bin/clang++ cmake -DMIOPEN_USE_COMGR=On -DBUILD_DEV=On -DCMAKE_BUILD_TYPE=debug -DMIOPEN_GPU_SYNC=On -DMIOPEN_TEST_FLAGS=--disable-verification-cache`

Test command
`CTEST_PARALLEL_LEVEL=4 MIOPEN_DEBUG_IMPLICIT_GEMM_NON_XDLOPS_INLINE_ASM=0 MIOPEN_CONV_PRECISE_ROCBLAS_TIMING=0 make -j64 check`

In my local machine yielded result of
`100% tests passed, 0 tests failed out of 55`

File: driver/conv_driver.hpp
Comment (atamazov) [inline driver/conv_driver.hpp:540]: What about 3D convolutions? "nchw" is inappropriate, for example.

File: src/include/miopen/conv/problem_description.hpp
Comment (atamazov) [inline src/include/miopen/conv/problem_description.hpp:170]: Why this change?

File: driver/conv_driver.hpp
Comment (jerryyin) [reply driver/conv_driver.hpp:540]: The in_layout option allows user to use any string as layout. So if it needs to be 3D convolution "nchwd", for example, user can set it. Setting it to "nchw" "nkhw" "kcyx" just so that it got defaulted to a set of coherent characters.

Once we agreed on `miopenSet4dTensorDescriptorLayout` interface that allows user to set the layout string field of a tensor, I can update the interface such that:

- Drop the `4d` in its name
- Formalize the set of characters and allowed length for layout string

File: src/include/miopen/conv/problem_description.hpp
Comment (jerryyin) [reply src/include/miopen/conv/problem_description.hpp:170]: The is related with the previous change: since input layout get defaulted to the lower case "nchw", the comparison is using "nchw". This can be addressed by, say, we formalized the layout string to be all lower or upper case.

Though I don't think I fully understand why it was originally written to compare with the upper cased "NCHW". Since `GetInLayout()` already get hard-coded to `NCHW` what's the point of the comparison here? Is it aiming for some future non-packed data?

File: src/conv/problem_description.cpp
Comment (jerryyin) [inline src/conv/problem_description.cpp:78]: TODO: reenable this. Any objection for me to adopt the approach from `Serialize()` function?
`F`/`B`/`W` or `forward`/`backwarddata`/`backwardweights`

File: src/include/miopen/conv/problem_description.hpp
Comment (atamazov) [inline src/include/miopen/conv/problem_description.hpp:164]: This function is used in `ProblemDescription::Serialize` that is used to generate database keys. Changing from "NCHW" to "nchw" would break find-db and, most likely, perf-db functionality.

File: src/conv/problem_description.cpp
Comment (atamazov) [inline src/conv/problem_description.cpp:110]: We must maintain this function backward compatible as it generates database keys. When layouts are default (NCHW-NKHW-KCYX) the modified function must return literally the same key as it does now.

/cc @JehandadKhan as this requirement is also important for Tuna, I think.

File: src/conv/problem_description.cpp
Comment (atamazov) [reply src/conv/problem_description.cpp:78]: Most likely , this is harmless, but I am not sure. Please discuss this with @DrizztDoUrden

File: src/include/miopen/conv/problem_description.hpp
Comment (jerryyin) [reply src/include/miopen/conv/problem_description.hpp:164]: Sounds good thanks for reminding. Looks like to be consistent I will have to adopt all upper letters for layouts.

File: src/conv/problem_description.cpp
Comment (jerryyin) [reply src/conv/problem_description.cpp:110]: Thanks. Will branch according to whether it is default layouts.

File: src/include/miopen/conv/problem_description.hpp
Comment (atamazov) [reply src/include/miopen/conv/problem_description.hpp:170]: > since input layout get defaulted to the lower case "nchw"

This change is undesired, please see https://github.com/ROCmSoftwarePlatform/MIOpen/pull/343#discussion_r456054013.

> Though I don't think I fully understand why it was originally written to compare with the upper cased "NCHW".

I am far from even moderate understanding. All this is legacy stuff, survived 4 years of refactoring and other changes. Originated from https://github.com/AMDComputeLibraries/MLOpen/pull/5/files#r456075197 4 years ago. It looks like dead code now.

And it will become a bug as soon as `GetInLayout()` will begin to return "NCDHW" for 3D tensors.

[Recommendation] Get rid of this in a separate PR.

Comment (jerryyin) [reply src/include/miopen/conv/problem_description.hpp:170]: Sounds good, will return here, but file a separate PR to return `GetInBatchSize() * GetInChannels() * GetInDepth() * GetInHeight() * GetInWidth() * GetInElementSize()` directly.

Comment (atamazov) [reply src/include/miopen/conv/problem_description.hpp:170]: My previous recommendation (getting rid of ternary) also applies to GetOutSize, GetWeightsSize

Comment (atamazov) [reply src/include/miopen/conv/problem_description.hpp:170]: > Sounds good, will revert here, but file a separate PR 

I recommend keeping this unchanged here, to avoid merge conflicts. Up to you of course.

File: src/conv/problem_description.cpp
Comment (JehandadKhan) [reply src/conv/problem_description.cpp:78]: We actually need it when parsing the database files to fix swapped elements by the problem descriptor.

Comment (jerryyin) [reply src/conv/problem_description.cpp]: Thanks for reminding, reverted to original 1/0

Comment (atamazov) [reply src/conv/problem_description.cpp]: > We actually need it when parsing the database files...

Database keys are produced by Serialize(), and direction is encoded there as F, B or W. NetworkConfig is used only in find-db, as a solution-specific, opaque payload.

**Comment by DrizztDoUrden** [comment]
Alternative would likely be to make conv ctx to reuse serialization from conv::ProblemDescription, but our current serialization process doesn't allow post-deserialization handling, so that won't be trivial.

File: src/conv/problem_description.cpp
Comment (DrizztDoUrden) [reply src/conv/problem_description.cpp]: I think network confing string representation is only used in conjunction with algo name string containing direction. But it would be better to check that, I am not constantly looking through every change to the library.

**Comment by jerryyin** [comment]
> Alternative would likely be to make conv ctx to reuse serialization from conv::ProblemDescription, but our current serialization process doesn't allow post-deserialization handling, so that won't be trivial.

No worries, It's been fairly straightforward to copy-o-past the solution to another place. I had it addressed in my last commit. I don't see a problem in the scope of this review. Though in the development process this has been very confusing and took me a while to realize why update in one problem description does not fix an issue I was having - Because there are two problem descriptions!

**Comment by DrizztDoUrden** [comment]
I ll have to clean up all the code to finally move to the new structure, but that got postponed for a while already.

File: src/conv/problem_description.cpp
Comment (JehandadKhan) [inline src/conv/problem_description.cpp:122]: For backwards compatibility the output layout should not be serialized either.

Comment (jerryyin) [reply src/conv/problem_description.cpp]: Thanks for catching this! Think I accidentally forgot about those two lines when adding the if condition. Fixed in d0f3423

**Comment by jerryyin** [comment]
Wasn't familiar with MIOpen development process, so gentle ping here. This PR has sit still for a while, what does it take to proceed with approval/merge? Is it blocked by CI failures? Let me know how to re-kick the CI for a new run.

**Comment by daniellowell** [comment]
All please review this PR.

**Comment by daniellowell** [comment]
@atamazov Please review this PR.

**Comment by daniellowell** [comment]
@jerryyin Please merge the latest `develop` into your branch. This should resolve being stuck at the COMgr stage.

**Comment by jerryyin** [comment]
@daniellowell Thanks, rebased to latest state of develop branch.

File: src/include/miopen/tensor.hpp
Comment (pfultz2) [inline src/include/miopen/tensor.hpp:189]: We should not be storing the layout as a string. The layout comes from the order of strides in the tensor descriptor.

File: include/miopen/miopen.h
Comment (pfultz2) [inline include/miopen/miopen.h:547]: We should not have this API. The layout can be set by using the strides or we can provide setters for the common layouts, such as `miopenSetCNHWTensorDescriptor` or `miopenSetNHWCTensorDescriptor`.

File: src/include/miopen/tensor.hpp
Comment (pfultz2) [inline src/include/miopen/tensor.hpp:177]: This is wrong as well. The layout should come from the order of the strides.

File: include/miopen/miopen.h
Comment (daniellowell) [reply include/miopen/miopen.h:547]: I don't understand the underlying reason why this should not be an API. The other options are good sure, but what is the actual objection?

Comment (pfultz2) [reply include/miopen/miopen.h:547]: The main problem is we have two ways to represent the layout. We can already specify CNHW layout with `miopenSetTensorDescriptor(tensorDesc, dataType, 4, {N, C, H, W}.begin(), {H*W, N*H*W, W, 1}.begin())`. Although this may not always be the most convenient for some users, which is why we should provide `miopenSetCNHWTensorDescriptor`. 

Using a string for the layout leads to us erroneously using the string to determine the layout instead of using the actual layout of the tensor. We could make the layout string and the actual layout of the tensor work, but this will double our test coverage. 

Also, MIGraphX will most likely specify the layout using the strides and not a layout string as it would be too cumbersome to infer.

File: src/include/miopen/tensor.hpp
Comment (pfultz2) [reply src/include/miopen/tensor.hpp:177]: Actually, the labels should be passed to `GetLayout` since the labels depend on the input type and not on the tensor layout. This can be implemented as:

```cpp
template <class Vector, class Op>
inline std::vector<int64_t> sort_permutation(const Vector& data, Op op)
{
    std::vector<std::int64_t> result(data.size());
    std::iota(result.begin(), result.end(), 0);
    std::sort(result.begin(), result.end(), [&](auto x, auto y) { return op(data[x], data[y]); });
    return result;
}

std::string TensorDescriptor::GetLayout(std::string labels) const
{
    auto result = labels;
    auto p = sort_permutation(strides, std::greater<>{});
    std::tranform(p.begin(), p.end(), result.begin(), [&](auto i) {
        return labels[i];
    });
    return result;
}
```

So then you can call it with `GetLayout("NCHW")` and if the layout is CNHW it will CNHW.

File: include/miopen/miopen.h
Comment (atamazov) [reply include/miopen/miopen.h:547]: +1 to Paul, I do not like strings here due to many reasons.

However I do not understand how `miopenSetTensorDescriptor(tensorDesc, dataType, 4, {N, C, H, W}.begin(), {H*W, N*H*W, W, 1}.begin())` can do the job. This doesn't look like a C code to me. Perhaps `{N, C, H, W}.begin()` and `{H*W, N*H*W, W, 1}.begin())` represent beginnings of buffers, but I do not see how the library can derive information about layout from two pointers. Maybe I just need to sleep some time? ;)

**Comment by atamazov** [comment]
@jerryyin 
> Wasn't familiar with MIOpen development process, so gentle ping here. This PR has sit still for a while, what does it take to proceed with approval/merge? Is it blocked by CI failures?

We often postpone review until all CI failures fixed, because fixing the CI problems may require significant code changes and reviewers will have repeat their work after that.

Please look at CI failures. If your PR has nothing to do with them, please inform us and explicitly ask for reviewing.

Please also tell how you've tested your PR (in the topmost comment). For example, often I pull and re-test branches during a review process.

File: include/miopen/miopen.h
Comment (pfultz2) [reply include/miopen/miopen.h:547]: >  This doesn't look like a C code to me.

This is not C code its C++ code. The `{N, C, H, W}` creates an initializer list and `.begin()` returns a pointer to it, so it can be passed to the C API. I was just trying to shorten it to one line. Another way to write it would be(assuming `N`, `C`, `H`, and `W` are constants):

```cpp
std::vector<int> lens = {N, C, H, W};
std::vector<int> strides = {H*W, N*H*W, W, 1};
miopenSetTensorDescriptor(tensorDesc, dataType, 4, lens.data(), strides.data());
```

> but I do not see how the library can derive information about layout from two pointers.

It derives the layout from the strides. As I showed below with the `GetLayout` function, `sort_permutation` will return the permutation `0,1,2,3` for NCHW layout and in the example above for CNHW it will return the permutation `1,0,2,3`.

If we pass the labels in as a string to `GetLayout` for the default permutation(ie `0,1,2,3`) then it will put the labels in the order of its layout. So `GetLayout("NCHW")` will return `NCHW` with permutation of `0,1,2,3` and it will return `CNHW` for permutation of `1,0,2,3`.

Comment (jerryyin) [reply include/miopen/miopen.h:547]: Thanks for the comments guys. It makes total sense after looking through especially @pfultz2 comments and suggestions. Using strides to infer the layout does make it less ambiguous and make it able to take advantage the existing API as well. Will push a new commit to address this.

File: src/include/miopen/tensor.hpp
Comment (jerryyin) [reply src/include/miopen/tensor.hpp:177]: Nice, it gives client the flexibility to specify the layout character strings too. I like this implementation.

**Comment by jerryyin** [comment]
@atamazov Thanks for reminding. We stand on the same page that reviewers should start reviewing only after fixing CI failures. Somehow I had the mis-understanding that CI wasn't fully functional due to `COMGR issues` but apparent it was fixed a while ago. 

Regarding with testing, it is less straightforward for this PR. I kept an `igemm-upstream` branch that had the integration of MIOpen and MLIR done. I was able to test a number of MIOpenDriver configs in different layouts in the context of that branch and things has been working fine. (No tuning related testing on my side, that might be how I missed the existing CI failures.) 

Considering that pushing all the changes in that branch at one go is messy and makes a huge reviewer effort, I decided to split that into a number of vertical PRs, just so that it is easier to review. Having said that, I'd be happy to share the instructions of how to test in the context of that branch, however, it might be too much to ask for a reviewer to do. What I will probably do is to also pull in the minimal amount of changes in `conv_driver` to make the PR test-able in its own context.

File: include/miopen/miopen.h
Comment (atamazov) [reply include/miopen/miopen.h]: @pfultz2 Thanks! it became clear to me in the morning, indeed :facepalm:

**Comment by jerryyin** [comment]
Please hold off review this PR. WIP to make CI happy.

**Comment by jerryyin** [comment]
Help needed: I can't figure out what make CI to consistently fail on two targets (in `test_conv2d` test case):
 - Hip clang debug target, in this [run](http://micimaster.amd.com/blue/organizations/jenkins/MLLibs%2FMIOpen/detail/zyin-layouts/16/pipeline)
 - gfx908 Hip debug target, in another [run](http://micimaster.amd.com/blue/organizations/jenkins/MLLibs%2FMIOpen/detail/zyin-layouts/17/pipeline)

For the first one, I have build the docker container from scratch and followed exactly on how CI builds and tests and all tests can pass just fine. For the second one, I tried to follow the process in a gfx906 but it always failed on building hcc. The ROCm version on the bare metal is 3.7 and is up to date. I don't know how to proceed further at this point.

File: src/conv/problem_description.cpp
Comment (jerryyin) [inline src/conv/problem_description.cpp:88]: TODO: remove the leftover from rebase

**Comment by jerryyin** [comment]
CI has a full pass, [here](http://micimaster.amd.com/blue/organizations/jenkins/MLLibs%2FMIOpen/detail/zyin-layouts/28/pipeline).

All please continue the review. Thanks.

**Comment by jerryyin** [comment]
Gentle ping @daniellowell @pfultz2. Let me know how to proceed with this PR.

**Comment by daniellowell** [comment]
@atamazov @JehandadKhan Changes have been made please re-review.

**Review by DrizztDoUrden — DISMISSED** [review]

File: src/include/miopen/tensor.hpp
Comment (atamazov) [inline src/include/miopen/tensor.hpp:188]: This involves creation/deletion of several containers, fillng, sorting etc. Not that fast as a simple getter should be. Shall we rename it to `ComputeLayout()`?

Comment (atamazov) [inline src/include/miopen/tensor.hpp:188]: Another consideration. When we support different layouts, layout comparisons (e.g. in `aSolver::IsApplicable()`) will happen quite often. Getting layouts is not very fast. Comparison of strings too. This could lead to some perf drop. Do you have a plan of optimization? Ideas?

File: src/include/miopen/tensor.hpp
Comment (atamazov) [inline src/include/miopen/tensor.hpp:196]: It seems like assignment is useless.

**Review by atamazov — CHANGES_REQUESTED** [review]
(Sorry for delaying the review)

File: driver/conv_driver.hpp
Comment (jerryyin) [reply driver/conv_driver.hpp:536]: Please refer to [this comment](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/343#discussion_r474778988). Because of the reason there I decided that updating layout string to include non-NCHW is orthogonal to this review. Plus I don't feel strong about using the similar set of letters being a big issue here.

File: src/solver/conv_MP_bidirectional_winograd.cpp
Comment (jerryyin) [reply src/solver/conv_MP_bidirectional_winograd.cpp:204]: Sounds reasonable. Though I'd want more clarification in this check: In the original code, why is only `weights_layout` checked here (instead of three)?

I'm asking this because most probably `params.IsLayoutDefault()` will check all three layout on whether they are default, then it becomes an un-equivalent code change compared to original one.

File: src/include/miopen/tensor.hpp
Comment (jerryyin) [reply src/include/miopen/tensor.hpp:188]: I see this in two ways:

- It does get more expensive, but in very limited manner. The 4-letter layout string is very easy to manipulate and adds only linear time complexity here. I don't think this is a big problem leaving it as-is.

- On the other hand, we can certainly cache it to a member variable (such that only the first time `GetLayout()` called that it is actually computed, otherwise it is just returning cached copies.). I can certainly implement this. Though I want to make sure you guys are okay with it in the sense that:
  1. It stores duplicated information against strides, as pointed out from original review comments.
  2. It is not very "functional", per say.

File: src/include/miopen/tensor.hpp
Comment (jerryyin) [reply src/include/miopen/tensor.hpp:196]: This is the solution I copy-pasted from Paul's implementation. It copy constructed the `result` from `lables` (instead the alternative is to `push_back()` in the transform stage), I assume it is slightly faster than the alternative.

File: src/include/miopen/tensor.hpp
Comment (jerryyin) [reply src/include/miopen/tensor.hpp:188]: We can decide on the name depends on how next review comments goes. Should we keep the function untouched, I shall rename it to `Compute`. Should we decide to cache the result, I will keep as is.

Comment (pfultz2) [reply src/include/miopen/tensor.hpp:188]: I think renaming it to `ComputeLayout` makes sense since it needs to do a caculation.

Comment (pfultz2) [reply src/include/miopen/tensor.hpp:188]: > On the other hand, we can certainly cache it to a member variable (such that only the first time GetLayout() called that it is actually computed, otherwise it is just returning cached copies.). 

I dont think this is a good idea. `GetLayout`(or `ComputeLayout`) should be `const`. 

It would be better to store the layout string in the `ConvolutionContext` so that it only needs to be computed once.

File: src/include/miopen/tensor.hpp
Comment (pfultz2) [reply src/include/miopen/tensor.hpp:196]: Yes, this is to avoid calling `push_back` as incrementing a pointer should be faster.

File: src/include/miopen/tensor.hpp
Comment (pfultz2) [inline src/include/miopen/tensor.hpp:179]: This can be declared as `static`.

File: src/solver/conv_MP_bidirectional_winograd.cpp
Comment (atamazov) [reply src/solver/conv_MP_bidirectional_winograd.cpp:204]: > why is only weights_layout checked here (instead of three)?

Because one line of useless code (comment) is enough to remind ;)

Comment (atamazov) [reply src/solver/conv_MP_bidirectional_winograd.cpp:204]: > then it becomes an un-equivalent code change compared to original one.

Original code does not check anything ;)

File: src/include/miopen/tensor.hpp
Comment (jerryyin) [reply src/include/miopen/tensor.hpp:188]: @pfultz2 Yah, think we mean the same, wasn't very clear about where to cache the layout string in my original comments. 

Addressed in latest commit.

File: src/include/miopen/tensor.hpp
Comment (jerryyin) [reply src/include/miopen/tensor.hpp]: Done. Thanks

File: src/include/miopen/tensor.hpp
Comment (jerryyin) [reply src/include/miopen/tensor.hpp:188]: Done, now I have both `ComputeLayout` and `GetLayout`.

**Comment by jerryyin** [comment]
FYI, the CI has some quite inconsistent result over the same code base again. I'm not sure its something that has been noticed. For example, an earlier run [(32) here](http://micimaster.amd.com/blue/organizations/jenkins/MLLibs%2FMIOpen/detail/zyin-layouts/32/pipeline/) complains that ROCm is not available on `gfx908 Hip debug`'s docker building step. It looks clear to me that different CI machines are configured differently which caused this inconsistency. My pr is hitting on this multiple times.

**Comment by daniellowell** [comment]
@jerryyin We had 2 bad nodes in our CI. I've disabled them.

**Comment by jerryyin** [comment]
Jenkins all passing [here](http://micimaster.amd.com/blue/organizations/jenkins/MLLibs%2FMIOpen/detail/zyin-layouts/40/pipeline). Let me know if you guys, @atamazov @pfultz2 have any further feedbacks. Thanks.

File: src/include/miopen/tensor.hpp
Comment (atamazov) [reply src/include/miopen/tensor.hpp:196]: Please add a comment that explains this. Otherwise the code looks like an error.

Comment (jerryyin) [reply src/include/miopen/tensor.hpp:198]: Done in f58abdaa7aa8b69d840066b9b20cae8ef58dc75c

File: driver/conv_driver.hpp
Comment (atamazov) [reply driver/conv_driver.hpp:536]: This seems Ok as an intermediate solution. However, what is your future plans about this? I mean support for independent layouts for input, output and weights. Right now these defaults looks confusing.

I suggest opening a dedicated ticket, possibly labeled `request_for_comments`, for discussion and making/documenting design decisions.

Comment (jerryyin) [reply driver/conv_driver.hpp:536]: I'm fine with adding this as an issue but not leading this effort. I'm willing to review and provide my feedback if there is a proposed solution, but I don't know what that would be at the moment. The primary reason has been twofold:

- I have very little idea about the history and previous design decisions for all the layout related implementation. This makes me a very poor owner for making a fundamental change to this solution, just because of how widely `buffer_info.hpp` has been included in MIOpen. I could be making (or propose) a dozen of changes not foreseeing the impact to existing code base. Therefore I'd like to scope my effort to the code base that is strictly igemm infrastructure where I could offer the most help.
- From the igemm perspective, I don't mind the actual string representation of layout - any 4-letter string will do the job. This does not impact the functionality of MIOpen existing code base - making it less of a priority. Readability-wise, it is not as good as NCHW-KCYX-NKHW, but each layout does come closely with a tensor so wouldn't be tricky to handle.

Comment (atamazov) [reply driver/conv_driver.hpp:536]: > I have very little idea about the history and previous design decisions

Nobody has I guess ;) I would say "it just happened".

> I could be making (or propose) a dozen of changes not foreseeing the impact to existing code base. 

That is why the reviewers are here. Working together, we all learn.

> Therefore I'd like to scope my effort to the code base that is strictly igemm infrastructure where  I could offer the most help.

I'll keep in mind this. However, as you can see, this is often impossible.

#509 created.

**Comment by atamazov** [comment]
@DrizztDoUrden Could you please have a look at this? It looks acceptable to me, but perhaps you have some substantial comments.

File: driver/conv_driver.hpp
Comment (jerryyin) [reply driver/conv_driver.hpp:536]: Appreciate the feedback and for helping creating this issue. It is unfortunate that I still have couple of changes pending and I'm trying to make more progress in this. I'm trying to keep focused and have it wrapped up sooner rather than later as maintaining a working feature branch is also non-trivial. To give more details here, I'd prioritize on a functional igemm code generation, just so as to unblock the tuning implementation for igemm (Adding layout for all tensors will impact perfdb and tuna, another collaboration point with MIOpen developers.)

**Comment by daniellowell** [comment]
@atamazov Your review is blocking this PR. Are there still outstanding issues?

**Comment by daniellowell** [comment]
@JehandadKhan  If the changes have resolved your issues, please unblock this PR.

**Review by JehandadKhan — APPROVED** [review]
This PR can go in, in a subsequent PR i would add support for perf db to support these changes.

**Review by atamazov — APPROVED** [review]
Tested for performance. No regressions. LGTM.

**Review by pfultz2 — APPROVED** [review]

**Pull request merged**
